src/apps/carbon-defi/ethereum/carbon-defi.strategy.contract-position-fetcher.ts

```diff
@@ -0,0 +1,142 @@
+  }
+
+  override async getDataProps(params: GetDataPropsParams<CarbonController, StrategyProps, StrategyDefinition>) {
+    return { strategy: params.definition.strategy };
+  }
+
+  override async filterPositionsForAddress(
+    address: string,
+    positions: ContractPosition<StrategyProps>[],
+  ): Promise<ContractPosition<StrategyProps>[]> {
+    return positions.filter(position => {
+      const owner = position.dataProps.strategy.owner;
+      return owner.toLowerCase() === address.toLowerCase();
+    });
+  }
+
+  // Override the getBalance query to keep token with budget set to 0
+  override async getBalances(address: string): Promise<ContractPositionBalance<StrategyProps>[]> {
+    const multicall = this.appToolkit.getViemMulticall(this.network);
+    if (address === ZERO_ADDRESS) return [];
+
+    const contractPositions = await this.getPositionsForBalances();
+    const filteredPositions = await this.filterPositionsForAddress(address, contractPositions);
+
+    const balances = await Promise.all(
+      filteredPositions.map(async contractPosition => {
+        const contract = multicall.wrap(this.getContract());
+        const balancesRaw = await this.getTokenBalancesPerPosition({ address, contract, contractPosition, multicall });
+        const tokens = contractPosition.tokens.map((cp, idx) =>
+          drillBalance(cp, balancesRaw[idx]?.toString() ?? '0', { isDebt: cp.metaType === MetaType.BORROWED }),
+        );
+        return {
+          ...contractPosition,
+          balanceUSD: sumBy(tokens, t => t.balanceUSD),
+          tokens,
+        };
+      }),
+    );
+
+    return balances;
+  }
+
+  async getLabel(params: GetDisplayPropsParams<CarbonController>): Promise<string> {
+    const tokens = params.contractPosition.tokens;
+    return `${tokens[0].symbol} / ${tokens[1].symbol}`;
+  }
+
+  async getDefinitions(): Promise<StrategyDefinition[]> {
+    const contract = this.getContract();
+    const address = contract.address;
+    const pairs = await contract.read.pairs();
+    const getStrategies = pairs.map(pair => contract.read.strategiesByPair([...pair, BigInt(0), BigInt(0)]));
+    const strategies = await Promise.all(getStrategies).then(matrix => matrix.flat());
+    return strategies.filter(isActiveStrategy).map(strategy => ({ address, strategy: strategy }));
+  }
+
+  async getTokenDefinitions(
+    params: GetTokenDefinitionsParams<CarbonController, StrategyDefinition>,
+  ): Promise<UnderlyingTokenDefinition[] | null> {
+    return params.definition.strategy.tokens.map(address => ({
+      address: address.toLowerCase().replace(ETH_ADDR_ALIAS, ZERO_ADDRESS),
+      metaType: MetaType.SUPPLIED,
+      network: this.network,
+    }));
+  }
+
+  async getTokenBalancesPerPosition(
+    params: GetTokenBalancesParams<CarbonController, StrategyProps>,
+  ): Promise<BigNumberish[]> {
+    const strategy = params.contractPosition.dataProps.strategy;
+    const [buy, sell] = strategy.orders;
+    return [buy.y, sell.y];
```

Comment:
You shouldn't propage the balances from the definiton => dataprops => balances
You should define positions in `GetDefinitions` and balances should be fetched within `getTokenBalancesPerPosition`

Reply:
The contract returns the balance alongside the tokens addresses anyway. Since `getDefinitions` is called before `getTokenBalancesPerPosition` I thought the `dataProps` was a good place to store it.
Another option is to use the `key` of the position to store the strategyID and call the contract for each position. But performance wise it's going to slow down the query.
Let me know what you prefer.